src/mlpack/methods/neighbor_search/ns_model.hpp

```diff
@@ -37,6 +49,121 @@ struct NSModelName<FurthestNeighborSort>
   static const std::string Name() { return "furthest_neighbor_search_model"; }
 };
 
+class SearchKVisitor : public boost::static_visitor<void>
+{
+ private:
+   const size_t k;
+   arma::Mat<size_t>& neighbors;
+   arma::mat& distances;
+
+ public:
+   template<typename NSType>
+   void operator()(NSType *ns) const;
+
+   SearchKVisitor(const size_t k,
+                  arma::Mat<size_t>& neighbors,
+                  arma::mat& distances);
+};
+
+class SearchVisitor : public boost::static_visitor<void>
```

Comment:
Maybe we should change the name of this class and `SearchKVisitor` to `BichromaticSearchVisitor` and `MonochromaticSearchVisitor`, respectively, for clarity?  I couldn't think of any shorter names to differentiate there, but maybe you have a better idea?  The root of my concern is that `Search` and `SearchK` don't really describe the difference in what is going on.